tests/BaseV3.php

```diff
@@ -0,0 +1,264 @@
+<?php
+
+namespace Tests;
+
+use PHPUnit\Framework\TestCase;
+
+ini_set('display_errors', 1);
+ini_set('display_startup_errors', 1);
+error_reporting(E_ALL);
+
+abstract class BaseV3 extends TestCase
+{
+    public function setUp(): void
+    {
+    }
+
+    public function tearDown(): void
+    {
+    }
+
+    private function execute($body = '', $url = '/', $method = 'POST', $headers = []) {
+        $ch = \curl_init();
+
+        $headers = \array_merge([
+            'content-type' => 'text/plain',
+            'x-open-runtimes-secret' => \getenv('OPEN_RUNTIMES_SECRET')
+        ], $headers);
+        $headersParsed = [];
+
+        foreach ($headers as $header => $value) {
+            $headersParsed[] = $header . ': ' . $value;
+        }
+
+        $responseHeaders = [];
+        $optArray = array(
+            CURLOPT_URL => 'http://localhost:3000' . $url,
+            CURLOPT_RETURNTRANSFER => true,
+            CURLOPT_HEADERFUNCTION => function ($curl, $header) use (&$responseHeaders) {
+                $len = strlen($header);
+                $header = explode(':', $header, 2);
+                if (count($header) < 2) // ignore invalid headers
+                    return $len;
+        
+                $key = strtolower(trim($header[0]));
+                $responseHeaders[$key] = trim($header[1]);
+
+                if(\in_array($key, ['x-open-runtimes-logs', 'x-open-runtimes-errors'])) {
+                    $responseHeaders[$key] = \urldecode($responseHeaders[$key]);
+                }
+        
+                return $len;
+            },
+            CURLOPT_CUSTOMREQUEST => $method,
+            CURLOPT_POSTFIELDS => \is_array($body) ? \json_encode($body, JSON_FORCE_OBJECT) : $body,
+            CURLOPT_HEADEROPT => \CURLHEADER_UNIFIED,
+            CURLOPT_HTTPHEADER => $headersParsed
+        );
+        
+        \curl_setopt_array($ch, $optArray);
+
+        $body = curl_exec($ch);
+        $code = curl_getinfo($ch, \CURLINFO_HTTP_CODE);
+
+        \curl_close($ch);
+
+        return [
+            'code' => $code,
+            'body' => $body,
+            'headers' => $responseHeaders
+        ];
+    }
+
+    public function testPlaintextResponse(): void
+    {
+        $response = $this->execute(headers: ['x-action' => 'plaintextResponse']);
+        self::assertEquals(200, $response['code']);
+        self::assertEquals('Hello World 👋', $response['body']);
+    }
+
+    public function testJsonResponse(): void
+    {
+        $response = $this->execute(headers: ['x-action' => 'jsonResponse']);
+        self::assertEquals(200, $response['code']);
```

Comment:
Hmm how is this test passing ? Looking at the code, a 204 response is returned

Reply:
Ignore 204 you noticed earlier. It was left-over code from earlier iteration that never occured. Such scehario was for not returning at all, which now results in 500 response tested here: https://github.com/open-runtimes/open-runtimes/blob/d7549b672133696085ac26e4c3da71877cecca12/tests/BaseV3.php#L108-L114